src/k2/dto/Expression.h

```diff
@@ -0,0 +1,106 @@
+#include <k2/transport/Payload.h>
+#include <k2/common/Common.h>
+#include <k2/transport/PayloadSerialization.h>
+#include <vector>
+#include <functional>
+#include "FieldTypes.h"
+#include "SKVRecord.h"
+
+namespace k2 {
+namespace dto {
+namespace expression {
+
+struct InvalidExpressionException : public std::exception {
+    virtual const char* what() const noexcept override{ return "InvalidExpression";}
+};
+
+struct TypeMismatchException : public std::exception {
+    virtual const char* what() const noexcept override { return "TypeMismatch"; }
+};
+
+enum class Operation : uint8_t {
+    EQ,           // value operands only
+    GT,           // value operands only
+    GTE,          // value operands only
+    LT,           // value operands only
+    LTE,          // value operands only
+    IS_NULL,      // value operands only
+    IS_TYPE,      // value operands only
+    STARTS_WITH,  // value operands only
+    CONTAINS,     // value operands only
+    ENDS_WITH,    // value operands only
+    AND,          // mixed value and/or nested expression operands
+    OR,           // mixed value and/or nested expression operands
+    XOR,          // mixed value and/or nested expression operands
+    NOT,           // mixed value and/or nested expression operands
+    UNKNOWN
+};
+
+// A Value in the expression model. It can be either
+// - a field reference which sets the fieldName, or
+// - a literal which is a user-supplied value (in the Payload literal) and type (in type).
+// If fieldName is non-empty, this is a reference. Otherwise it is a literal value
+struct Value {
+    String fieldName;
+    FieldType type = FieldType::NOT_KNOWN;
+    Payload literal;
+    K2_PAYLOAD_FIELDS(fieldName, type, literal);
+};
+
+// An Expression in the expression model.
+// The operation is applied in the order that the children are in the vector. So a binary
+// operator like LT would be applied as valueChildren[0] < valueChildren[1]
+struct Expression {
+    Operation op = Operation::UNKNOWN;
+    std::vector<Expression> expressionChildren;
+    std::vector<Value> valueChildren;
```

Comment:
Should we consider using SschematizeValue directly, i.e. Id + field Idx instead of fieldName in valueChildren of Expression? Since when the client construct the col/field reference Value, it has already resolved the column/field name with schema, so using fieldName still will be less efficient as it need to be resolved again.

Reply:
The issue is that the schematized value is computed on each record separately on the server side. It has to be done this way for reference values. I was thinking of adding some optimization later on for literals, but since these need to be sent over the wire, the optimization has to be done again on the server side. I was thinking of some lazy caching for these literals.